flang/lib/Lower/Bridge.cpp

```diff
@@ -2006,19 +2013,39 @@ class FirConverter : public Fortran::lower::AbstractConverter {
           auto dim = builder->createIntegerConstant(loc, idxTy, iter.index());
           dimInfo = builder->create<fir::BoxDimsOp>(loc, idxTy, idxTy, idxTy,
                                                     box, dim);
-          extents.emplace_back(dimInfo.getResult(1));
+          mlir::Value lb;
           if (auto low = spec->lbound().GetExplicit()) {
             auto expr = Fortran::semantics::SomeExpr{*low};
-            auto lb =
-                builder->createConvert(loc, idxTy, createFIRExpr(loc, &expr));
-            lbounds.emplace_back(lb);
+            lb = builder->createConvert(loc, idxTy, createFIRExpr(loc, &expr));
           } else {
             // FIXME: The front-end is not setting up the implicit lower
             // bounds to 1 for assumed shape array. Do this here for now,
             // but that is absolutely wrong for allocatable and pointers.
-            // lbounds.emplace_back(dimInfo.getResult(0));
-            lbounds.emplace_back(builder->createIntegerConstant(loc, idxTy, 1));
+            // lb = dimInfo.getResult(0);
+            lb = builder->createIntegerConstant(loc, idxTy, 1);
           }
+          lbounds.emplace_back(lb);
+          // FIXME: there is a division by zero risk, and this is wrong in case
+          // the input first stride is not a multiple of the element size (may
+          // happen when passing an array of subcomponent e.eg
+          // foo(aDerivedArray%someComponent).) Having a ByteStrideOp would be
+          // nicer...
+          auto step = builder->create<mlir::SignedDivIOp>(
+              loc, idxTy, dimInfo.getResult(2), elementSize);
```

Comment:
Hmm. Are we really requiring byte strides when constructing the FIR? That seems like a oversight, more than a feature.

Lowering to a SliceOp should be easier than this, I think.

Reply:
Yeah, it's kind a ugly, but this is not a fir design issue, the code here is dealing with assumed shape on the callee side, the issue here is we are trying "lifting" the stride form a "fir.box" to a SliceOp.  The third field is the stride in bytes as per Fortran 2018 18.5.2, so fir design of fir.box is correct in this regard. And SliceOp design also looks good to me when it comes to lower actual slices.

The lowering of a slice to a SclieOp (when we lower a user slice like A(i:j:k)) is done here, and is indeed very simple:
https://github.com/flang-compiler/f18-llvm-project/pull/542/files#diff-6bb3b4929e13ba83349cd5d5e97eabaf0d3b88a87b83710c3842c69ca8801a6fR1113-R1131